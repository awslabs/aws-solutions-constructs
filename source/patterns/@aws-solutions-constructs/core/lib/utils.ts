/**
 *  Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import * as deepmerge from 'deepmerge';
import { flagOverriddenDefaults } from './override-warning-service';
import * as log from 'npmlog';

function isObject(val: object) {
    return val != null && typeof val === 'object'
        && Object.prototype.toString.call(val) === '[object Object]';
}

function isPlainObject(o: object) {
  if (Array.isArray(o) === true) {
    return true;
  }

  if (isObject(o) === false) {
    return false;
  }

  // If has modified constructor
  const ctor = o.constructor;
  if (typeof ctor !== 'function') {
    return false;
  }

  // If has modified prototype
  const prot = ctor.prototype;
  if (isObject(prot) === false) {
    return false;
  }

  // If constructor does not have an Object-specific method
  if (prot.hasOwnProperty('isPrototypeOf') === false) {
    return false;
  }

  // Most likely a plain Object
  return true;
}

function combineMerge(target: any[], source: any[]) {
    return target.concat(source);
}

function overwriteMerge(target: any[], source: any[]) {
  target = source;
  return target;
}

export function overrideProps(DefaultProps: object, userProps: object, concatArray: boolean = false): any {
  // Notify the user via console output if defaults are overridden
  const overrideWarningsEnabled = (process.env.overrideWarningsEnabled !== 'false');
  if (overrideWarningsEnabled) {
    flagOverriddenDefaults(DefaultProps, userProps);
  }
  // Override the sensible defaults with user provided props
  if (concatArray) {
    return deepmerge(DefaultProps, userProps, {
      arrayMerge: combineMerge,
      isMergeableObject: isPlainObject
    });
  } else {
    return deepmerge(DefaultProps, userProps, {
      arrayMerge: overwriteMerge,
      isMergeableObject: isPlainObject
    });
  }
}

export function printWarning(message: string) {
  // Style the log output
  log.prefixStyle.bold = true;
  log.prefixStyle.fg = 'red';
  log.enableColor();
  log.warn('AWS_SOLUTIONS_CONSTRUCTS_WARNING: ', message);
}
