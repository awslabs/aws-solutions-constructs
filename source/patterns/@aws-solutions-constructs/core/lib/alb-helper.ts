/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

/*
 *  The functions found here in the core library are for internal use and can be changed
 *  or removed outside of a major release. We recommend against calling them directly from client code.
 */

import * as elb from "aws-cdk-lib/aws-elasticloadbalancingv2";
import { Construct } from "constructs";
import * as ec2 from "aws-cdk-lib/aws-ec2";
import * as s3 from "aws-cdk-lib/aws-s3";
import * as ecs from "aws-cdk-lib/aws-ecs";
import * as lambda from "aws-cdk-lib/aws-lambda";
import { ApplicationProtocol, ListenerAction, } from "aws-cdk-lib/aws-elasticloadbalancingv2";
import * as elbt from "aws-cdk-lib/aws-elasticloadbalancingv2-targets";
import { printWarning, consolidateProps } from "./utils";
import { DefaultListenerProps } from "./alb-defaults";
import { createAlbLoggingBucket } from "./s3-bucket-helper";
import { DefaultLoggingBucketProps } from "./s3-bucket-defaults";

/**
 * @internal This is an internal core function and should not be called directly by Solutions Constructs clients.
 *
 * Returns the correct ALB Load Balancer to use in this construct, either an existing
 * one provided as an argument or create  new one otherwise.
 */
export function ObtainAlb(
  scope: Construct,
  id: string,
  vpc: ec2.IVpc,
  publicApi: boolean,
  existingLoadBalancerObj?: elb.ApplicationLoadBalancer,
  loadBalancerProps?: elb.ApplicationLoadBalancerProps | any,
  logAccessLogs?: boolean,
  loggingBucketProps?: s3.BucketProps
): elb.ApplicationLoadBalancer {
  let loadBalancer: elb.ApplicationLoadBalancer;

  if (existingLoadBalancerObj) {
    loadBalancer = existingLoadBalancerObj;
  } else {
    const consolidatedProps = consolidateProps({}, loadBalancerProps, { vpc, internetFacing: publicApi });
    loadBalancer = new elb.ApplicationLoadBalancer(
      scope,
      `${id}-alb`,
      consolidatedProps
    );
    if (logAccessLogs === undefined || logAccessLogs === true) {
      const consolidatedLoggingBucketProps = consolidateProps(DefaultLoggingBucketProps(), loggingBucketProps);
      const loggingBucket = createAlbLoggingBucket(scope, id, consolidatedLoggingBucketProps);
      loadBalancer.logAccessLogs(loggingBucket);
    }
  }
  return loadBalancer;
}

/**
 * @internal This is an internal core function and should not be called directly by Solutions Constructs clients.
 */
export function AddListener(
  scope: Construct,
  id: string,
  loadBalancer: elb.ApplicationLoadBalancer,
  listenerProps: elb.ApplicationListenerProps | any
): elb.ApplicationListener {
  let consolidatedListenerProps: elb.ApplicationListenerProps;

  consolidatedListenerProps = consolidateProps(DefaultListenerProps(loadBalancer), listenerProps);

  //  create the listener
  const listener = new elb.ApplicationListener(
    scope,
    `${id}-listener`,
    consolidatedListenerProps
  );
  loadBalancer.listeners.push(listener);

  if (consolidatedListenerProps.protocol === elb.ApplicationProtocol.HTTP) {
    // This will use core.printWarning in the actual construct
    printWarning(
      "AWS recommends encrypting traffic to an Application Load Balancer using HTTPS."
    );
    if (listenerProps.certificates?.length > 0) {
      throw new Error("HTTP listeners cannot use a certificate");
    }
  } else {
    if (!listenerProps.certificates || listenerProps.certificates.length === 0) {
      throw new Error("A listener using HTTPS protocol requires a certificate");
    }

    listener.addCertificates(`${id}-cert`, listenerProps.certificates);
  }

  if (consolidatedListenerProps.protocol === elb.ApplicationProtocol.HTTPS) {
    const opt: elb.RedirectOptions = {
      port: "443",
      protocol: "HTTPS",
    };

    const httpListener = new elb.ApplicationListener(
      scope,
      `${id}-redirect`,
      {
        loadBalancer,
        protocol: ApplicationProtocol.HTTP,
        defaultAction: ListenerAction.redirect(opt),
      }
    );
    loadBalancer.listeners.push(httpListener);
  }

  return listener;
}

/**
 * @internal This is an internal core function and should not be called directly by Solutions Constructs clients.
 *
 * Creates a Target Group for Lambda functions and adds the
 * provided functions as a target to that group. Then adds
 * the new Target Group to the provided Listener.
 */
export function AddLambdaTarget(
  scope: Construct,
  id: string,
  currentListener: elb.ApplicationListener,
  lambdaFunction: lambda.IFunction,
  ruleProps?: elb.AddRuleProps,
  targetProps?: elb.ApplicationTargetGroupProps,
): elb.ApplicationTargetGroup  {

  //  Create the target and assign it to a new target group
  const lambdaTarget = new elbt.LambdaTarget(lambdaFunction);
  const newTargetGroup = new elb.ApplicationTargetGroup(scope, `${id}-tg`, {
    targets: [lambdaTarget],
    targetGroupName: targetProps ? targetProps.targetGroupName : undefined,
    healthCheck: targetProps ? targetProps.healthCheck : undefined
  });

  // The interface AddRuleProps includes conditions and priority, combine that
  // with targetGroups and we can assemble AddApplicationTargetGroupProps
  const consolidatedTargetProps = consolidateProps({}, ruleProps, { targetGroups: [newTargetGroup] });
  currentListener.addTargetGroups(`${scope.node.id}-targets`, consolidatedTargetProps);

  newTargetGroup.setAttribute('stickiness.enabled', undefined);
  return newTargetGroup;
}

/**
 * @internal This is an internal core function and should not be called directly by Solutions Constructs clients.
 */
export function AddFargateTarget(
  scope: Construct,
  id: string,
  currentListener: elb.ApplicationListener,
  fargateService: ecs.FargateService,
  ruleProps?: elb.AddRuleProps,
  targetProps?: elb.ApplicationTargetGroupProps,
): elb.ApplicationTargetGroup  {

  if (targetProps?.protocol !== elb.ApplicationProtocol.HTTPS) {
    printWarning('AWS recommends using HTTPS protocol for Target Groups in production applications');
  }

  const newTargetGroup = new elb.ApplicationTargetGroup(scope, `${id}-tg`, targetProps);

  // The interface AddRuleProps includes conditions and priority, combine that
  // with targetGroups and we can assemble an AddApplicationTargetGroupProps object
  const consolidatedTargetProps = consolidateProps({ targetGroups: [newTargetGroup] }, ruleProps);

  currentListener.addTargetGroups(`${scope.node.id}-targets`, consolidatedTargetProps);
  newTargetGroup.addTarget(fargateService);

  return newTargetGroup;
}

/**
 * @internal This is an internal core function and should not be called directly by Solutions Constructs clients.
 *
 * Looks for the listener associated with Target Groups
 * If there is a single listener, this returns it whether it is HTTP or HTTPS
 * If there are 2 listeners, it finds the HTTPS listener (we assume the HTTP listener redirects to HTTPS)
 */
export function GetActiveListener(listeners: elb.ApplicationListener[]): elb.ApplicationListener {
  let listener: elb.ApplicationListener;

  if (listeners.length === 0 ) {
    throw new Error(`There are no listeners in the ALB`);
  }
  if (listeners.length === 1 ) {
    listener = listeners[0];
  } else {
    listener = listeners.find(i => (i.node.children[0] as elb.CfnListener).protocol === "HTTPS") as elb.ApplicationListener;
  }
  return listener;
}

/**
 * @internal This is an internal core function and should not be called directly by Solutions Constructs clients.
 */
export function CheckAlbProps(props: any) {
  let errorMessages = '';
  let errorFound = false;

  if (props.listenerProps?.certificateArns) {
    errorMessages += "certificateArns is deprecated. Please supply certificates using props.listenerProps.certificates\n";
    errorFound = true;
  }

  if (
    ((props.existingLoadBalancerObj &&
      props.existingLoadBalancerObj.listeners.length === 0) ||
      !props.existingLoadBalancerObj) &&
    !props.listenerProps
  ) {
    errorMessages += "When adding the first listener and target to a load balancer, listenerProps must be specified and include at least a certificate or protocol: HTTP\n";
    errorFound = true;
  }

  if (
    props.existingLoadBalancerObj &&
    props.existingLoadBalancerObj.listeners.length > 0 &&
    props.listenerProps
  ) {
    errorFound = true;
    errorMessages += "This load balancer already has a listener, listenerProps may not be specified\n";
  }

  if (
    props.existingLoadBalancerObj &&
    props.existingLoadBalancerObj.listeners.length > 0 &&
    !props.ruleProps
  ) {
    errorFound = true;
    errorMessages += "When adding a second target to an existing listener, there must be rules provided\n";
  }

  // Check construct specific invalid inputs
  if (props.existingLoadBalancerObj && !props.existingVpc) {
    errorFound = true;
    errorMessages += "An existing ALB is already in a VPC, that VPC must be provided in props.existingVpc for the rest of the construct to use.\n";
  }

  if (props.loadBalancerProps?.vpc) {
    errorFound = true;
    errorMessages += 'Specify any existing VPC at the construct level, not within loadBalancerProps.\n';
  }

  if (props.existingLoadBalancerObj) {
    printWarning(
      "The public/private property of an existing ALB must match the props.publicApi setting provided."
    );
  }

  if (errorFound) {
    throw new Error(errorMessages);
  }
}