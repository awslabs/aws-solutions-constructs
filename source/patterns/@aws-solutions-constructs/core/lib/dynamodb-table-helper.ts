/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

/*
 *  The functions found here in the core library are for internal use and can be changed
 *  or removed outside of a major release. We recommend against calling them directly from client code.
 */

import * as dynamodb from 'aws-cdk-lib/aws-dynamodb';
import { DefaultTableProps, DefaultTableWithStreamProps } from './dynamodb-table-defaults';
import { consolidateProps } from './utils';
// Note: To ensure CDKv2 compatibility, keep the import statement for Construct separate
import { Construct } from 'constructs';

export interface BuildDynamoDBTableProps {
  /**
   * Optional user provided props to override the default props
   *
   * @default - Default props are used
   */
  readonly dynamoTableProps?: dynamodb.TableProps,
  /**
   * Existing instance of dynamodb table object.
   * Providing both this and `dynamoTableProps` will cause an error.
   *
   * @default - None
   */
  readonly existingTableObj?: dynamodb.Table
  /**
   * Existing instance of dynamodb interface.
   * Providing both this and `dynamoTableProps` will cause an error.
   *
   * @default - None
   */
  readonly existingTableInterface?: dynamodb.ITable
}

export interface BuildDynamoDBTableWithStreamProps {
  /**
   * Optional user provided props to override the default props
   *
   * @default - Default props are used
   */
  readonly dynamoTableProps?: dynamodb.TableProps,
  /**
   * Existing instance of dynamodb table object.
   * Providing both this and `dynamoTableProps` will cause an error.
   *
   * @default - None
   */
  readonly existingTableInterface?: dynamodb.ITable
}

export interface BuildDynamoDBTableResponse {
  readonly tableInterface: dynamodb.ITable,
  readonly tableObject?: dynamodb.Table,
}

/**
 * @internal This is an internal core function and should not be called directly by Solutions Constructs clients.
 */
export function buildDynamoDBTable(scope: Construct, props: BuildDynamoDBTableProps): BuildDynamoDBTableResponse {
  checkTableProps(props);

  // Conditional DynamoDB Table creation
  if (props.existingTableObj) {
    return { tableInterface: props.existingTableObj, tableObject: props.existingTableObj };
  } else if (props.existingTableInterface) {
    return { tableInterface: props.existingTableInterface };
  } else {
    const consolidatedTableProps = consolidateProps(DefaultTableProps, props.dynamoTableProps);
    const newTable = new dynamodb.Table(scope, 'DynamoTable', consolidatedTableProps);
    return { tableInterface: newTable, tableObject: newTable };
  }
}

/**
 * @internal This is an internal core function and should not be called directly by Solutions Constructs clients.
 */
export function checkTableProps(props: BuildDynamoDBTableProps) {
  let errorMessages = '';
  let errorFound = false;

  if (props.dynamoTableProps && props.existingTableObj) {
    errorMessages += 'Error - Either provide existingTableObj or dynamoTableProps, but not both.\n';
    errorFound = true;
  }

  if (props.dynamoTableProps && props.existingTableInterface) {
    errorMessages += 'Error - Either provide existingTableInterface or dynamoTableProps, but not both.\n';
    errorFound = true;
  }

  if (props.existingTableObj && props.existingTableInterface) {
    errorMessages += 'Error - Either provide existingTableInterface or existingTableObj, but not both.\n';
    errorFound = true;
  }

  if (errorFound) {
    throw new Error(errorMessages);
  }
}

export interface BuildDynamoDBTableWithStreamResponse {
  readonly tableInterface: dynamodb.ITable,
  readonly tableObject?: dynamodb.Table,
}

/**
 * @internal This is an internal core function and should not be called directly by Solutions Constructs clients.
 */
export function buildDynamoDBTableWithStream(scope: Construct, props: BuildDynamoDBTableWithStreamProps): BuildDynamoDBTableWithStreamResponse {
  // Conditional DynamoDB Table creation
  if (!props.existingTableInterface) {
    // Set the default props for DynamoDB table
    const dynamoTableProps = consolidateProps(DefaultTableWithStreamProps, props.dynamoTableProps);
    const dynamoTable: dynamodb.Table = new dynamodb.Table(scope, 'DynamoTable', dynamoTableProps);
    return { tableInterface: dynamoTable, tableObject: dynamoTable };
  } else {
    return { tableInterface: props.existingTableInterface };
  }
}

/**
 * @internal This is an internal core function and should not be called directly by Solutions Constructs clients.
 */
export function getPartitionKeyNameFromTable(table: dynamodb.Table): string {
  const cfnTable = table.node.findChild('Resource') as dynamodb.CfnTable;
  const keySchema = cfnTable.keySchema as dynamodb.CfnTable.KeySchemaProperty[];
  const partitionKey = keySchema.find((keyPart: any) => keyPart.keyType === 'HASH');
  if (!partitionKey) {
    throw new Error('Partition key for table not defined');
  }
  return partitionKey.attributeName;
}
