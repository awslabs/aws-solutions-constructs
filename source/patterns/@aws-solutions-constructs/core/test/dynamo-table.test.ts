/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import { Stack } from 'aws-cdk-lib';
import * as dynamodb from 'aws-cdk-lib/aws-dynamodb';
import * as defaults from '../index';
import { overrideProps } from '../lib/utils';
import { Template } from 'aws-cdk-lib/assertions';
import { getPartitionKeyNameFromTable } from '../lib/dynamodb-table-helper';

test('test TableProps change billing mode', () => {
  const stack = new Stack();

  const defaultProps: dynamodb.TableProps = defaults.DefaultTableProps;

  const inProps: dynamodb.TableProps = {
    billingMode: dynamodb.BillingMode.PROVISIONED,
    readCapacity: 3,
    writeCapacity: 3,
    partitionKey: {
      name: 'id',
      type: dynamodb.AttributeType.STRING
    }
  };

  const outProps = overrideProps(defaultProps, inProps);
  new dynamodb.Table(stack, 'test-dynamo-override', outProps);

  const template = Template.fromStack(stack);
  template.hasResourceProperties("AWS::DynamoDB::Table", {
    KeySchema: [
      {
        AttributeName: "id",
        KeyType: "HASH"
      }
    ],
    AttributeDefinitions: [
      {
        AttributeName: "id",
        AttributeType: "S"
      }
    ],
    ProvisionedThroughput: {
      ReadCapacityUnits: 3,
      WriteCapacityUnits: 3
    },
    SSESpecification: {
      SSEEnabled: true
    }
  });
});

test('test TableProps override add sort key', () => {
  const stack = new Stack();

  const defaultProps: dynamodb.TableProps = defaults.DefaultTableProps;

  const inProps: dynamodb.TableProps = {
    partitionKey: {
      name: 'id',
      type: dynamodb.AttributeType.STRING
    },
    sortKey: {
      name: 'sort_key',
      type: dynamodb.AttributeType.STRING
    }
  };

  const outProps = overrideProps(defaultProps, inProps);
  new dynamodb.Table(stack, 'test-dynamo-override', outProps);

  const template = Template.fromStack(stack);
  template.hasResourceProperties("AWS::DynamoDB::Table", {
    KeySchema: [
      {
        AttributeName: "id",
        KeyType: "HASH"
      },
      {
        AttributeName: "sort_key",
        KeyType: "RANGE"
      }
    ],
    AttributeDefinitions: [
      {
        AttributeName: "id",
        AttributeType: "S"
      },
      {
        AttributeName: "sort_key",
        AttributeType: "S"
      }
    ],
    BillingMode: "PAY_PER_REQUEST",
    SSESpecification: {
      SSEEnabled: true
    }
  });
});

test('test TableWithStreamProps override stream view type', () => {
  const stack = new Stack();

  const defaultProps: dynamodb.TableProps = defaults.DefaultTableWithStreamProps;

  const inProps: dynamodb.TableProps = {
    partitionKey: {
      name: 'id',
      type: dynamodb.AttributeType.STRING
    },
    stream: dynamodb.StreamViewType.NEW_IMAGE
  };

  const outProps = overrideProps(defaultProps, inProps);
  new dynamodb.Table(stack, 'test-dynamo-override', outProps);

  const template = Template.fromStack(stack);
  template.hasResourceProperties("AWS::DynamoDB::Table", {
    KeySchema: [
      {
        AttributeName: "id",
        KeyType: "HASH"
      }
    ],
    AttributeDefinitions: [
      {
        AttributeName: "id",
        AttributeType: "S"
      }
    ],
    BillingMode: "PAY_PER_REQUEST",
    SSESpecification: {
      SSEEnabled: true
    },
    StreamSpecification: {
      StreamViewType: "NEW_IMAGE"
    }
  });
});

test('test buildDynamoDBTable with existingTableObj', () => {
  const stack = new Stack();

  const tableProps: dynamodb.TableProps = {
    billingMode: dynamodb.BillingMode.PROVISIONED,
    partitionKey: {
      name: 'table_id',
      type: dynamodb.AttributeType.STRING
    }
  };

  const existingTableObj = new dynamodb.Table(stack, 'DynamoTable', tableProps);

  const buildDynamoDBTableResponse = defaults.buildDynamoDBTable(stack, {
    existingTableObj
  });

  expect(buildDynamoDBTableResponse.tableInterface).toBeDefined();
  expect(buildDynamoDBTableResponse.tableObject).toBeDefined();

  const template = Template.fromStack(stack);
  template.hasResourceProperties('AWS::DynamoDB::Table', {
    KeySchema: [
      {
        AttributeName: "table_id",
        KeyType: "HASH"
      }
    ]
  });

  template.hasResourceProperties('AWS::DynamoDB::Table', {
    ProvisionedThroughput: {
      ReadCapacityUnits: 5,
      WriteCapacityUnits: 5
    }
  });
});

test('test buildDynamoDBTable without any arguments', () => {
  const stack = new Stack();

  const buildDynamoDBTableResponse = defaults.buildDynamoDBTable(stack, {});

  expect(buildDynamoDBTableResponse.tableInterface).toBeDefined();
  expect(buildDynamoDBTableResponse.tableObject).toBeDefined();

  const template = Template.fromStack(stack);
  template.hasResourceProperties('AWS::DynamoDB::Table', {
    KeySchema: [
      {
        AttributeName: "id",
        KeyType: "HASH"
      }
    ]
  });

  template.hasResourceProperties('AWS::DynamoDB::Table', {
    BillingMode: "PAY_PER_REQUEST"
  });

  template.hasResourceProperties('AWS::DynamoDB::Table', {
    SSESpecification: {
      SSEEnabled: true
    }
  });
});

test('test buildDynamoDBTable with TableProps', () => {
  const stack = new Stack();

  const dynamoTableProps: dynamodb.TableProps = {
    billingMode: dynamodb.BillingMode.PROVISIONED,
    partitionKey: {
      name: 'table_id',
      type: dynamodb.AttributeType.STRING
    }
  };

  const buildDynamoDBTableResponse = defaults.buildDynamoDBTable(stack, {
    dynamoTableProps
  });

  expect(buildDynamoDBTableResponse.tableInterface).toBeDefined();
  expect(buildDynamoDBTableResponse.tableObject).toBeDefined();

  const template = Template.fromStack(stack);
  template.hasResourceProperties('AWS::DynamoDB::Table', {
    KeySchema: [
      {
        AttributeName: "table_id",
        KeyType: "HASH"
      }
    ]
  });

  template.hasResourceProperties('AWS::DynamoDB::Table', {
    ProvisionedThroughput: {
      ReadCapacityUnits: 5,
      WriteCapacityUnits: 5
    }
  });
});

test('test buildDynamoDBTableWithStream with TableProps', () => {
  const stack = new Stack();

  const dynamoTableProps: dynamodb.TableProps = {
    partitionKey: {
      name: 'table_id',
      type: dynamodb.AttributeType.STRING
    },
    stream: dynamodb.StreamViewType.NEW_IMAGE
  };

  const response = defaults.buildDynamoDBTableWithStream(stack, {
    dynamoTableProps
  });

  expect(response.tableInterface).toBeDefined();
  expect(response.tableObject).toBeDefined();

  const template = Template.fromStack(stack);
  template.hasResourceProperties('AWS::DynamoDB::Table', {
    KeySchema: [
      {
        AttributeName: "table_id",
        KeyType: "HASH"
      }
    ]
  });

  template.hasResourceProperties('AWS::DynamoDB::Table', {
    StreamSpecification: {
      StreamViewType: "NEW_IMAGE"
    }
  });
});

test('test buildDynamoDBTableWithStream without any arguments', () => {
  const stack = new Stack();

  const response = defaults.buildDynamoDBTableWithStream(stack, {});

  expect(response.tableInterface).toBeDefined();
  expect(response.tableObject).toBeDefined();

  const template = Template.fromStack(stack);
  template.hasResourceProperties('AWS::DynamoDB::Table', {
    KeySchema: [
      {
        AttributeName: "id",
        KeyType: "HASH"
      }
    ]
  });

  template.hasResourceProperties('AWS::DynamoDB::Table', {
    BillingMode: "PAY_PER_REQUEST"
  });

  template.hasResourceProperties('AWS::DynamoDB::Table', {
    SSESpecification: {
      SSEEnabled: true
    }
  });

  template.hasResourceProperties('AWS::DynamoDB::Table', {
    StreamSpecification: {
      StreamViewType: "NEW_AND_OLD_IMAGES"
    }
  });
});

test('test buildDynamoDBTableWithStream with existingTableObj', () => {
  const stack = new Stack();

  const tableProps: dynamodb.TableProps = {
    partitionKey: {
      name: 'table_id',
      type: dynamodb.AttributeType.STRING
    },
    stream: dynamodb.StreamViewType.NEW_IMAGE
  };

  const existingTableInterface = new dynamodb.Table(stack, 'DynamoTable', tableProps);

  const response = defaults.buildDynamoDBTableWithStream(stack, {
    existingTableInterface
  });

  expect(response.tableInterface).toBeDefined();
  expect(response.tableObject).not.toBeDefined();

  const template = Template.fromStack(stack);
  template.hasResourceProperties('AWS::DynamoDB::Table', {
    KeySchema: [
      {
        AttributeName: "table_id",
        KeyType: "HASH"
      }
    ]
  });

  template.hasResourceProperties('AWS::DynamoDB::Table', {
    StreamSpecification: {
      StreamViewType: "NEW_IMAGE"
    }
  });
});

test('test buildDynamoDBTable with existingTableInterface', () => {
  const stack = new Stack();

  const tableProps: dynamodb.TableProps = {
    partitionKey: {
      name: 'table_id',
      type: dynamodb.AttributeType.STRING
    },
    stream: dynamodb.StreamViewType.NEW_IMAGE
  };

  const existingTableInterface = new dynamodb.Table(stack, 'DynamoTable', tableProps);

  const buildDynamoDBTableResponse = defaults.buildDynamoDBTable(stack, {
    existingTableInterface
  });

  expect(buildDynamoDBTableResponse.tableInterface).toBeDefined();
  expect(buildDynamoDBTableResponse.tableObject).not.toBeDefined();

  const template = Template.fromStack(stack);
  template.hasResourceProperties('AWS::DynamoDB::Table', {
    KeySchema: [
      {
        AttributeName: "table_id",
        KeyType: "HASH"
      }
    ]
  });

  template.hasResourceProperties('AWS::DynamoDB::Table', {
    StreamSpecification: {
      StreamViewType: "NEW_IMAGE"
    }
  });
});

test('test getPartitionKeyNameFromTable()', () => {
  const partitionKeyName = 'testPartitionKey';

  const stack = new Stack();

  const defaultProps: dynamodb.TableProps = defaults.DefaultTableProps;

  const inProps: dynamodb.TableProps = {
    partitionKey: {
      name: partitionKeyName,
      type: dynamodb.AttributeType.STRING
    },
    sortKey: {
      name: 'sort_key',
      type: dynamodb.AttributeType.STRING
    }
  };

  const outProps = overrideProps(defaultProps, inProps);
  const newTable = new dynamodb.Table(stack, 'test-dynamo-override', outProps);

  const testKeyName = getPartitionKeyNameFromTable(newTable);

  expect(testKeyName).toEqual(partitionKeyName);
});

test('Test providing both existingTableInterface and existingTableObj', () => {
  const stack = new Stack();

  const tableProps: dynamodb.TableProps = {
    partitionKey: {
      name: 'table_id',
      type: dynamodb.AttributeType.STRING
    },
    stream: dynamodb.StreamViewType.NEW_IMAGE
  };

  const existingTableInterface = new dynamodb.Table(stack, 'DynamoTable', tableProps)
  ;
  const newProps = {
    existingTableInterface,
    existingTableObj: existingTableInterface
  };
  const app = () => {
    defaults.buildDynamoDBTable(stack, newProps);
  };

  expect(app).toThrowError('Error - Either provide existingTableInterface or existingTableObj, but not both.\n');
});

test('Test providing both existingTableInterface and dynamoTableProps', () => {
  const stack = new Stack();

  const dynamoTableProps: dynamodb.TableProps = {
    partitionKey: {
      name: 'table_id',
      type: dynamodb.AttributeType.STRING
    },
    stream: dynamodb.StreamViewType.NEW_IMAGE
  };

  const existingTableInterface = new dynamodb.Table(stack, 'DynamoTable', dynamoTableProps)
  ;
  const newProps = {
    existingTableInterface,
    dynamoTableProps
  };
  const app = () => {
    defaults.buildDynamoDBTable(stack, newProps);
  };

  expect(app).toThrowError('Error - Either provide existingTableInterface or dynamoTableProps, but not both.\n');
});

test('Test providing both existingTableObj and dynamoTableProps', () => {
  const stack = new Stack();

  const dynamoTableProps: dynamodb.TableProps = {
    partitionKey: {
      name: 'table_id',
      type: dynamodb.AttributeType.STRING
    },
    stream: dynamodb.StreamViewType.NEW_IMAGE
  };

  const existingTableObj = new dynamodb.Table(stack, 'DynamoTable', dynamoTableProps)
  ;
  const newProps = {
    existingTableObj,
    dynamoTableProps
  };
  const app = () => {
    defaults.buildDynamoDBTable(stack, newProps);
  };

  expect(app).toThrowError('Error - Either provide existingTableObj or dynamoTableProps, but not both.\n');
});