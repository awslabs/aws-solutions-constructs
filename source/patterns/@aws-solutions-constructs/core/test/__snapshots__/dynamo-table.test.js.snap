// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot test TableProps default params 1`] = `
Object {
  "Resources": Object {
    "testdynamodefaults72AF3E8C": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "AttributeDefinitions": Array [
          Object {
            "AttributeName": "id",
            "AttributeType": "S",
          },
        ],
        "BillingMode": "PAY_PER_REQUEST",
        "KeySchema": Array [
          Object {
            "AttributeName": "id",
            "KeyType": "HASH",
          },
        ],
        "PointInTimeRecoverySpecification": Object {
          "PointInTimeRecoveryEnabled": true,
        },
        "SSESpecification": Object {
          "SSEEnabled": true,
        },
      },
      "Type": "AWS::DynamoDB::Table",
      "UpdateReplacePolicy": "Retain",
    },
  },
}
`;

exports[`snapshot test TableWithStream default params 1`] = `
Object {
  "Resources": Object {
    "testdynamostreamdefaultsFD08DF32": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "AttributeDefinitions": Array [
          Object {
            "AttributeName": "id",
            "AttributeType": "S",
          },
        ],
        "BillingMode": "PAY_PER_REQUEST",
        "KeySchema": Array [
          Object {
            "AttributeName": "id",
            "KeyType": "HASH",
          },
        ],
        "PointInTimeRecoverySpecification": Object {
          "PointInTimeRecoveryEnabled": true,
        },
        "SSESpecification": Object {
          "SSEEnabled": true,
        },
        "StreamSpecification": Object {
          "StreamViewType": "NEW_AND_OLD_IMAGES",
        },
      },
      "Type": "AWS::DynamoDB::Table",
      "UpdateReplacePolicy": "Retain",
    },
  },
}
`;
